kernel/src/arch/aarch64/pgt48.rs

```diff
@@ -0,0 +1,356 @@
+    }
+}
+
+impl PagingImpl for PageTable {
+    fn new<'alloc>(mm: &mut mm::MemoryManager) -> &'alloc mut Self {
+        // FIXME: No unwrap here
+        let page = mm.alloc_pages(1).unwrap();
+        let page_table: *mut PageTable = page.into();
+        // FIXME: Do not unwrap either
+        let page_table = unsafe { page_table.as_mut().unwrap() };
+
+        page_table
+            .entries
+            .iter_mut()
+            .for_each(|content| unsafe { &mut content.descriptor }.set_invalid());
+
+        page_table
+    }
+
+    fn get_page_size() -> usize {
+        4096
+    }
+
+    fn map(
+        &mut self,
+        mm: &mut mm::MemoryManager,
+        pa: mm::PAddr,
+        va: mm::VAddr,
+        perms: mm::Permissions,
+    ) -> Result<(), Error> {
+        self.map_inner(Some(mm), pa.into(), va.into(), perms)?;
+
+        Ok(())
+    }
+
+    fn map_noalloc(
+        &mut self,
+        pa: mm::PAddr,
+        va: mm::VAddr,
+        perms: mm::Permissions,
+    ) -> Result<(), paging::Error> {
+        self.map_inner(None, pa.into(), va.into(), perms)?;
+
+        Ok(())
+    }
+
+    fn add_invalid_entry(
+        &mut self,
+        mm: &mut mm::MemoryManager,
+        vaddr: mm::VAddr,
+    ) -> Result<(), Error> {
+        let mut entry = self.map_inner(
+            Some(mm),
+            PAddr(0x0A0A_0A0A_0A0A_0A0A),
+            vaddr.into(),
+            mm::Permissions::READ,
+        )?;
+
+        entry.set_invalid();
+
+        Ok(())
+    }
+
+    fn reload(&mut self) {
+        MAIR_EL1.write(
+            // Attribute 0 - NonCacheable normal DRAM. FIXME: enable cache?
+            MAIR_EL1::Attr0_Normal_Outer::NonCacheable + MAIR_EL1::Attr0_Normal_Inner::NonCacheable,
+        );
+        TTBR0_EL1.set_baddr((self as *const PageTable) as u64);
+        TCR_EL1.write(
+            TCR_EL1::TBI0::Used
+                + TCR_EL1::IPS::Bits_48
+                + TCR_EL1::TG0::KiB_4
+                // + TCR_EL1::SH0::Inner
+                + TCR_EL1::SH0::None
+                // + TCR_EL1::ORGN0::WriteBack_ReadAlloc_WriteAlloc_Cacheable
+                + TCR_EL1::ORGN0::NonCacheable
+                // + TCR_EL1::IRGN0::WriteBack_ReadAlloc_WriteAlloc_Cacheable
+                + TCR_EL1::IRGN0::NonCacheable
+                + TCR_EL1::EPD0::EnableTTBR0Walks
+                + TCR_EL1::A1::TTBR0
+                + TCR_EL1::T0SZ.val(16)
+                + TCR_EL1::EPD1::DisableTTBR1Walks,
+        );
+
+        unsafe {
+            barrier::isb(barrier::SY);
+        }
+
+        SCTLR_EL1.modify(SCTLR_EL1::M::Enable);
+
+        unsafe {
+            barrier::isb(barrier::SY);
+        }
+    }
+
+    fn disable(&mut self) {
+        // let satp = Satp::new()
+        //     .with_ppn(0)
+        //     .with_asid(0)
+        //     .with_mode(SatpMode::Bare as u8);
+
+        // unsafe {
+        //     asm!("csrw satp, {}", in(reg)u64::from(satp));
+        //     asm!("sfence.vma");
+        // }
```

Comment:
Dead code? Or is that the valid version but kept away for now behind the `unimplemented!()`?